File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.py
import sys
input = sys.stdin.readline
from collections import deque
n = int(input())

table = [list(map(int, input().strip())) for _ in range(n)]
sol = []
dx = [1, -1, 0, 0]
dy = [0, 0, 1, -1]
# print(*table, sep='\n')

def bfs(x, y):
    q = deque()
    q.append([x, y])
    table[y][x] = 0
    cnt = 1
    while q:
        x, y = q.popleft()
        # print(*table, sep='\n')
        # print()
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            if 0 <= nx < n and 0 <= ny < n and table[ny][nx] == 1:
                q.append([nx, ny])
                table[ny][nx] = 0
                cnt += 1
    sol.append(cnt)


for i in range(n):
    for j in range(n):
        if table[i][j] == 1:
            bfs(j, i)

# print(*table, sep='\n')

print(len(sol))
sol.sort()
for i in sol:
    print(i)


# for k in range(2, counter+1):
#     cnt = 0
#     for i in range(n):
#         for j in range(n):
#             if table[i][j] == k:
#                 cnt += 1
#     sol.append(cnt)
#
# sol.sort()
# for i in sol:
#     print(i)

File: 백준/Silver/2667. 단지번호붙이기/단지번호붙이기.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

class Pair {
    int x;
    int y;

    public Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
}

public class Main {

    static int n;
    static int[][] map;
    static boolean[][] visited;
    static int[] dx = {1, -1, 0, 0};
    static int[] dy = {0, 0, 1, -1};
    static List<Integer> listAns;


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st;
        n = Integer.parseInt(br.readLine());
        map = new int[n][n];
        visited = new boolean[n][n];
        listAns = new ArrayList<>();

        for (int i = 0; i < n; i++) {
            String[] tmp = br.readLine().split("");
            for (int j = 0; j < n; j++) {
                map[i][j] = Integer.parseInt(tmp[j]);
            }
        }

        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                if (map[i][j] == 1 && !visited[i][j]) {
                    listAns.add(bfs(i, j));
                }
            }
        }

        Collections.sort(listAns);

        System.out.println(listAns.size());
        for (int x : listAns) {
            System.out.println(x);
        }

        br.close();

    }

    static int bfs(int y, int x) {
        Queue<Pair> q = new LinkedList<>();
        int cnt = 1;

        q.add(new Pair(x, y));
        visited[y][x] = true;

        while (!q.isEmpty()) {
            Pair p = q.poll();

            for (int i = 0; i < 4; i++) {
                int nx = p.x + dx[i];
                int ny = p.y + dy[i];

                if (nx >= 0 && nx < n && ny >= 0 && ny < n && !visited[ny][nx] && map[ny][nx] != 0) {
                    q.add(new Pair(nx, ny));
                    visited[ny][nx] = true;
                    cnt++;
                }
            }

        }

        return cnt;

    }

}

